File: web/src/api/services/product.service.ts
import type { Page, Product, ProductPayload } from '#/types'
import { api } from '..'

class ProductService {
  private readonly basePath = '/products'

  async list(params: {
    page: number
    size: number
    search?: string
  }): Promise<Page<Product>> {
    const { data } = await api.get<Page<Product>>(this.basePath, { params })

    return data
  }

  async byId(id: number | string): Promise<Product> {
    const { data } = await api.get<Product>(`${this.basePath}/${id}`)

    return data
  }

  async create(body: ProductPayload): Promise<Product> {
    const { data } = await api.post<Product>(this.basePath, body)

    return data
  }

  async update(id: number | string, body: ProductPayload): Promise<Product> {
    const { data } = await api.put<Product>(`${this.basePath}/${id}`, body)

    return data
  }

  async remove(id: number | string): Promise<void> {
    await api.delete(`${this.basePath}/${id}`)
  }
}

export const productService = new ProductService()

File: server/src/main/java/com/quispe/food_delivery/product/ProductService.java
package com.quispe.food_delivery.product;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.quispe.food_delivery.common.exception.ResourceNotFoundException;
import com.quispe.food_delivery.product.dto.ProductRequestDto;
import com.quispe.food_delivery.product.dto.ProductResponseDto;
import com.quispe.food_delivery.product.entity.Product;
import com.quispe.food_delivery.product.mapper.ProductMapper;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class ProductService {

    private final ProductRepository productRepository;
    private final ProductMapper productMapper;

    @Transactional(readOnly = true)
    public Page<ProductResponseDto> findAll(Pageable pageable) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        boolean isAdmin = authentication != null && authentication.getAuthorities().stream()
                .anyMatch(a -> "ROLE_ADMIN".equals(a.getAuthority()));

        Page<Product> products = isAdmin
                ? productRepository.findAll(pageable)
                : productRepository.findAllByActiveTrue(pageable);

        return products.map(productMapper::toResponse);
    }

    @Transactional(readOnly = true)
    public ProductResponseDto findById(Long id) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Produto não encontrado com ID: " + id));
        return productMapper.toResponse(product);
    }

    @Transactional
    public ProductResponseDto create(ProductRequestDto request) {
        Product product = productMapper.toEntity(request);
        Product savedProduct = productRepository.save(product);
        return productMapper.toResponse(savedProduct);
    }

    @Transactional
    public ProductResponseDto update(Long id, ProductRequestDto request) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Produto não encontrado com ID: " + id));

        product.setName(request.getName());
        product.setDescription(request.getDescription());
        product.setPrice(request.getPrice());
        if (request.getActive() != null) {
            product.setActive(request.getActive());
        }

        Product updatedProduct = productRepository.save(product);
        return productMapper.toResponse(updatedProduct);
    }

    @Transactional
    public void delete(Long id) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Produto não encontrado com ID: " + id));
        product.setActive(false);
        productRepository.save(product);
    }
}
